File: 94.binary-tree-inorder-traversal.py
#
# @lc app=leetcode id=94 lang=python3
#
# [94] Binary Tree Inorder Traversal
#
from typing import List
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# @lc code=start
class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        # 중위순회
        # 스스로 풀기는 했는데 솔루션의 다른 방법은 안 해봤음

        if not root:
            return []

        history = []
        stack = [root]

        count = 0
        while stack:
            node = stack.pop()
            if node.left:
                if node.left in history:
                    history.append(node)
                    if node.right:
                        stack.append(node.right)
                else:
                    stack.append(node)
                    stack.append(node.left)
            else:
                history.append(node)
                if node.right:
                    stack.append(node.right)

        for i in range(len(history)):
            history[i] = history[i].val
        return history


# @lc code=end

File: Binary Tree/src/94.binary-tree-inorder-traversal.js
/*
 * @lc app=leetcode id=94 lang=javascript
 *
 * [94] Binary Tree Inorder Traversal
 */

//Definition for a binary tree node.
class TreeNode {
  constructor(val, left, right) {
     this.val = (val===undefined ? 0 : val)
     this.left = (left===undefined ? null : left)
     this.right = (right===undefined ? null : right)
  }
}
// @lc code=start
/**
 * @param {TreeNode} node
 * @param {number[]} stack
 * @return null
 */
const repeat = (node, stack) => {
  if (node.left !== null) {
    repeat(node.left, stack);
  }
  stack.push(node.val);

  if (node.right !== null) {
    repeat(node.right, stack);
  }
};
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
  if (root === null) {
    return [];
  }

  const stack = [];

  // Iteration
  // const answer = [];
  // let node = root;
  // while (stack.length > 0 || node !== null) {
  //   while (node !== null) {
  //     stack.push(node);
  //     node = node.left;
  //   }

  //   node = stack.pop();
  //   answer.push(node.val);
  //   node = node.right;
  // }
  // return answer;

  // Recursion
  repeat(root, stack);
  return stack;
};
// @lc code=end

